File: src/main/java/eu/mshade/enderframe/inventory/Inventory.java
package eu.mshade.enderframe.inventory;

import eu.mshade.enderframe.Agent;
import eu.mshade.enderframe.Watchable;
import eu.mshade.enderframe.item.ItemStack;
import eu.mshade.enderframe.item.Material;
import eu.mshade.enderframe.item.MaterialKey;
import eu.mshade.enderframe.mojang.chat.TextComponent;

import java.util.*;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.function.Consumer;
import java.util.function.Function;

public abstract class Inventory implements Watchable {

    protected InventoryKey inventoryKey;
    protected ItemStack[] itemStacks;
    protected UUID uniqueId;
    protected List<Integer> updateSlots = new ArrayList<>();
    protected Queue<Agent> agents = new ConcurrentLinkedQueue<>();

    public Inventory(InventoryKey inventoryKey, ItemStack[] itemStacks, UUID uniqueId) {
        this.inventoryKey = inventoryKey;
        this.uniqueId = uniqueId;
        this.itemStacks = itemStacks;
        InventoryRepository.INSTANCE.add(this);
    }

    public Inventory(InventoryKey inventoryKey, UUID uniqueId) {
        this(inventoryKey, new ItemStack[inventoryKey.getDefaultSlot()], uniqueId);
    }

    public InventoryKey getInventoryKey() {
        return inventoryKey;
    }

    public void setItemStack(int slot, ItemStack itemStack) {
        if (itemStack != null && itemStack.getMaterial() != null && !itemStack.getMaterial().equals(Material.AIR)) this.itemStacks[slot] = itemStack;
        else this.itemStacks[slot] = null;
        updateSlots.add(slot);
    }


    public void deleteItemStack(int slot){
        this.itemStacks[slot] = null;
        updateSlots.add(slot);
    }

    public ItemStack getItemStack(int slot){
        return this.itemStacks[slot];
    }

    public ItemStack[] getItemStacks() {
        return itemStacks;
    }

    public int findFirstEmptySlot(){
        return findFirstEmptySlot(0);
    }

    public int findFirstEmptySlot(int offset){
        for (int i = offset; i < itemStacks.length; i++) {
            ItemStack itemStack = getItemStack(i);
            if (itemStack == null) return i;
        }
        return -1;
    }

    public ItemStack findItemStack(MaterialKey materialKey, Function<ItemStack, Boolean> filter){
        return findItemStack(0, materialKey, filter);
    }

    public ItemStack findItemStack(int offset, MaterialKey materialKey, Function<ItemStack, Boolean> filter){
        for (int i = offset; i < itemStacks.length; i++) {
            ItemStack itemStack = getItemStack(i);
            if (itemStack == null || !itemStack.getMaterial().equals(materialKey)) continue;
            if (filter.apply(itemStack)) {
                return itemStack;
            }
        }
        return null;
    }


    public int getSize() {
        return itemStacks.length;
    }

    public void fill(ItemStack[] itemStacks) {
        for (int i = 0; i < itemStacks.length; i++) {
            setItemStack(i, itemStacks[i]);
        }
    }

    public UUID getUniqueId() {
        return uniqueId;
    }

    public Collection<Integer> visitUpdateSlots() {
        Collection<Integer> updateSlots = new ArrayList<>(this.updateSlots);
        this.updateSlots.clear();
        return updateSlots;
    }

    @Override
    public void addWatcher(Agent agent) {
        agents.add(agent);
    }

    @Override
    public void removeWatcher(Agent agent) {
        agents.remove(agent);
    }

    @Override
    public Collection<Agent> getWatchers() {
        return agents;
    }

    @Override
    public boolean isWatching(Agent agent) {
        return agents.contains(agent);
    }

    @Override
    public void notify(Consumer<Agent> agent) {
        agents.forEach(agent);
    }
}

File: src/main/java/eu/mshade/enderframe/inventory/Inventory.kt
package eu.mshade.enderframe.inventory

import eu.mshade.enderframe.Agent
import eu.mshade.enderframe.Watchable
import eu.mshade.enderframe.item.ItemStack
import eu.mshade.enderframe.item.Material
import eu.mshade.enderframe.item.MaterialKey
import java.util.*
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.function.Consumer
import java.util.function.Function

abstract class Inventory(var inventoryKey: InventoryKey, @JvmField var itemStacks: Array<ItemStack?>) : Watchable {
    protected var updateSlots: MutableList<Int> = ArrayList()
    protected var agents: Queue<Agent> = ConcurrentLinkedQueue()

    constructor(inventoryKey: InventoryKey) : this(
        inventoryKey,
        arrayOfNulls<ItemStack>(inventoryKey.defaultSlot),
    )

    open fun setItemStack(slot: Int, itemStack: ItemStack?) {
        if (itemStack != null && itemStack.material != null && itemStack.material != Material.AIR) itemStacks[slot] =
            itemStack else itemStacks[slot] = null
        updateSlots.add(slot)
    }

    open fun deleteItemStack(slot: Int) {
        itemStacks[slot] = null
        updateSlots.add(slot)
    }

    open fun getItemStack(slot: Int): ItemStack? {
        return itemStacks[slot]
    }

    open fun getItemStack(equipmentSlot: EquipmentSlot): ItemStack? {
        return getItemStack(equipmentSlot.ordinal)
    }

    open fun setItemStack(equipmentSlot: EquipmentSlot, itemStack: ItemStack?) {
        setItemStack(equipmentSlot.ordinal, itemStack)
    }

    open fun findFirstEmptySlot(): Int {
        return findFirstEmptySlot(0)
    }

    open fun findFirstEmptySlot(offset: Int): Int {
        for (i in offset until itemStacks.size) {
            val itemStack = getItemStack(i)
            if (itemStack == null || itemStack.amount < itemStack.material.maxStackSize) return i
        }
        return -1
    }

    fun addItemStack(itemStack: ItemStack): Boolean {
        val (slot, foundItemStack) = findItemStack(itemStack.material) { item -> item != null && item.matchMetadata(itemStack) && item.amount < item.material.maxStackSize }
        if (foundItemStack == null || foundItemStack.amount >= itemStack.material.maxStackSize) {
            val findFirstEmptySlot = findFirstEmptySlot()
            if (findFirstEmptySlot != -1) {
                setItemStack(findFirstEmptySlot, itemStack)
                updateSlots.add(findFirstEmptySlot)
                return true
            }
            return false
        }

        if (foundItemStack.amount + itemStack.amount > itemStack.material.maxStackSize) {
            val rest = (foundItemStack.amount + itemStack.amount) - itemStack.material.maxStackSize
            foundItemStack.amount = itemStack.material.maxStackSize
            updateSlots.add(slot)

            val findFirstEmptySlot = findFirstEmptySlot()
            if (findFirstEmptySlot != -1) {
                setItemStack(findFirstEmptySlot, itemStack.clone().apply { amount = rest })
                updateSlots.add(findFirstEmptySlot)
                return true
            }
            return false
        }else {
            foundItemStack.modifyAmount { amount -> itemStack.amount + amount }
            updateSlots.add(slot)
        }
        return true
    }

    fun findItemStack(materialKey: MaterialKey, filter: Function<ItemStack?, Boolean>): Pair<Int, ItemStack?> {
        return findItemStack(0, materialKey, filter)
    }

    fun findItemStack(offset: Int, materialKey: MaterialKey, filter: Function<ItemStack?, Boolean>): Pair<Int, ItemStack?> {
        for (i in offset until itemStacks.size) {
            val itemStack = getItemStack(i)
            if (itemStack == null || itemStack.material != materialKey) continue
            if (filter.apply(itemStack)) {
                return Pair(i, itemStack)
            }
        }
        return Pair(-1, null)
    }


    val size: Int
        get() = itemStacks.size
    val rows: Int
        get() = size / 9

    fun fill(itemStacks: Array<ItemStack?>) {
        for (i in itemStacks.indices) {
            setItemStack(i, itemStacks[i])
        }
    }


    fun visitUpdateSlots(): Collection<Int> {
        val updateSlots: Collection<Int> = ArrayList(updateSlots)
        this.updateSlots.clear()
        return updateSlots
    }

    fun isInUpdateSlots(slot: Int): Boolean {
        return updateSlots.contains(slot)
    }

    override fun addWatcher(agent: Agent) {
        agents.add(agent)
    }

    override fun removeWatcher(agent: Agent) {
        agents.remove(agent)
    }

    override fun getWatchers(): Collection<Agent> {
        return agents
    }

    override fun isWatching(agent: Agent): Boolean {
        return agents.contains(agent)
    }

}